rita_client/src/operator_update/mod.rs

```diff
@@ -168,48 +169,69 @@ pub async fn operator_update(ops_last_seen_usage_hour: Option<u64>) -> u64 {
+    let send_hours = current_hour - last_seen_hour > 1;
     let mut usage_tracker_data: Option<UsageTracker> = None;
-    // only deal with this if we actually do need to send some hours
     // if ops_last_seen_usage_hour is a None the thread has restarted and we are waiting for ops to tell us how much
-    // data we need to send, which will be populated with the next checkin cycle
-    if hours_to_send != 0 && ops_last_seen_usage_hour.is_some() {
+    // data we need to send, which will be populated with the next checkin cycle. 730 is the average numbers of hours
+    // in a month, and we only send 1 month at a time if ops is requesting the full usage history.
+    if send_hours && ops_last_seen_usage_hour.is_some() {
         let mut usage_data_client = get_usage_data(Client);
         let mut usage_data_relay = get_usage_data(Relay);
         let mut new_client_data: VecDeque<UsageHour> = VecDeque::new();
         let mut new_relay_data: VecDeque<UsageHour> = VecDeque::new();
-        for _hour in 0..hours_to_send {
-            // pop front, add to front of new vecdeque if exists.
-            while let Some(data) = usage_data_client.pop_front() {
-                if data.index > last_seen_hour {
+
+        // ops is expecting data as [oldest..newest] but now we sent newest .. oldest
+        let client_oldest = match usage_data_client.back() {
+            Some(x) => x.index,
+            None => 0,
+        };
+        let relay_oldest = match usage_data_relay.back() {
+            Some(x) => x.index,
+            None => 0,
+        };
+        // if the last seen hour is earlier (lower) than our earliest saved index, we are uploading entire history so don't worry about position since last seen
+        if last_seen_hour < client_oldest {
+            new_client_data = iterate_month_usage_data(usage_data_client);
+        } else if client_oldest == last_seen_hour + 2 {
+            // we are simply updating the second most recent (because the most recent may be incomplete until it rolls over)
+            if let Some(data) = usage_data_client.get(1) {
+                if data.index == last_seen_hour + 1 {
                     new_client_data.push_front(UsageHour {
                         up: data.up,
                         down: data.down,
                         price: data.price,
                         index: data.index,
                     });
-                    break;
                 }
             }
-
-            // pop front, add to front of new vecdeque if exists.
-            while let Some(data) = usage_data_relay.pop_front() {
-                if data.index > last_seen_hour {
+        } else {
+            // binary search until we find the last seen hour or hour in the vecdeque with index before that (as in cases of gaps in data)
+            let pos = find_position_since_last_seen(&usage_data_client, last_seen_hour);
```

Comment:
are we sure the array is sorted? you should probably call sort before you call this search (most sort algos are pretty efficient if the array is already sorted) 

Even if we *think* it's sorted on most routers I'm sure we'll have at least one weird one.

Also need a unit test that tries feeding in unsorted data just to be sure here.